### PR TITLE
🐛 Bug fix on java validator for failed test due to $DEFAULT at rule evaluation

### PR DESCRIPTION
**Pull request by GeorgeLuo**
business logic change + test for fix + update tagchowder to newer version

### CODE REVIEW

File: validator/java/pom.xml [hunk 1]
```diff
@@ -557,7 +557,7 @@
         <dependency>
             <groupId>com.yahoo.tagchowder</groupId>
             <artifactId>tagchowder.core</artifactId>
-            <version>2.0.10</version>
+            <version>2.0.17</version>
```
Comment (rsimha) [inline validator/java/pom.xml:560]: Thanks for the upgrade. Could we also fix this security vulnerability?

image

Comment (GeorgeLuo) [reply validator/java/pom.xml:560]: alright. There was some conflict with maven checkstyles, seems to be resolved, lets see if this builds.

Comment (GeorgeLuo) [reply validator/java/pom.xml:560]: whoops ignore this last commit. TIL 8.8 < 8.18

File: validator/java/src/main/java/dev/amp/validator/css/Canonicalizer.java [hunk 2]
```diff
@@ -52,7 +52,7 @@
     public Canonicalizer(@Nonnull final Map<String, ValidatorProtos.AtRuleSpec.BlockType> atRuleSpec,
                          @Nonnull final ValidatorProtos.AtRuleSpec.BlockType defaultSpec) {
         this.atRuleSpec = atRuleSpec;
-        this.defaultAtRuleSpec = defaultSpec;
+         this.defaultAtRuleSpec = defaultSpec;
```
Comment (rsimha) [inline validator/java/src/main/java/dev/amp/validator/css/Canonicalizer.java:55]: Revert?

File: validator/java/ampvalidator_checkstyle.xml [hunk 3]
```diff
@@ -9,8 +9,8 @@
         <!-- Make the @SuppressWarnings annotations available to Checkstyle -->
         <module name="SuppressWarningsHolder" />
         <module name="JavadocMethod">
-            <property name="suppressLoadErrors" value="true"/>
-            <property name="allowUndeclaredRTE" value="true"/>
+<!--            <property name="suppressLoadErrors" value="true"/>-->
+<!--            <property name="allowUndeclaredRTE" value="true"/>-->
```
Comment (rsimha) [inline validator/java/ampvalidator_checkstyle.xml:13]: Delete instead of comment?

**Comment by twifkak** [comment]
@Gregable This looks to be a port of cl/305733898 (#27784).

**Comment by honeybadgerdontcare** [comment]
> @Gregable This looks to be a port of cl/305733898 (#27784).

Yes. They need to get these engine changes in so the tests can pass for Java.

File: validator/java/src/main/java/dev/amp/validator/css/Canonicalizer.java (hunk 2 above)
Comment (GeorgeLuo) [reply validator/java/src/main/java/dev/amp/validator/css/Canonicalizer.java]: good eye

**Comment by rsimha** [comment]
@GeorgeLuo @honeybadgerdontcare @Gregable A couple questions about these tests:

1. From the latest [travis logs](https://travis-ci.org/github/ampproject/amphtml/jobs/676465659#L357), it appears the tests are still broken. Let me know if this isn't expected, and if you need help debugging.
2. Several of the dependencies in `validator/java/pom.xml` are out of date, and some of them are showing up as security vulnerabilities in our repo (see https://github.com/ampproject/amphtml/pull/27692#pullrequestreview-395661096). I've enabled auto-upgrades for this file using our existing package auto-updater, but the java validator failures are preventing us from merging all the [upgrade PRs](https://github.com/ampproject/amphtml/pulls?q=is%3Apr+is%3Aopen+%22validator%2Fjava%2Fpom.xml%22) (15 and counting). Once the tests themselves are fixed, we'll need to a) re-test and re-run all these upgrade PRs or b) do a one-time manual upgrade of all packages in `validator/java/pom.xml` and then let automated maintenance take over. Can you help with this?

**Comment by GeorgeLuo** [comment]
@rsimha yes, I can run a clean pull and see if tests fail, due to discrepancies with my local repos. I might ask to enable debugging for the build if it works from my local environment, just for a little visibility.

**Comment by rsimha** [comment]
> @rsimha yes, I can run a clean pull and see if tests fail, due to discrepancies with my local repos. I might ask to enable debugging for the build if it works from my local environment, just for a little visibility.

Thanks for the help. If you're developing on Linux, you should be able to run these tests by [checking out](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) the `amphtml` repo and invoking `gulp validator-java`:

For local debugging, you'll see more logging by changing `-verbose` from `1` to `2` here:

https://github.com/ampproject/amphtml/blob/62248e7ade786d78afa6fd8a0f24a6b3697d2afb/validator/java/BUILD#L74

**Comment by GeorgeLuo** [comment]
Thank you! My maven build with a clean env was successful. When trying with gulp it actually looks it's hanging after the resource files are pulled down, with "Did you forget to signal async completion".  It doesn't look like the tests are even ran. 

This happens in the case when the files already exist as well. Either a resolution comes from the gulp job or our shell script to fetch resources needs modification. Have you seen this class of error before?

**Comment by rsimha** [comment]
> Have you seen this class of error before?

I'm unfamiliar with this. Can you increase the verbosity in your PR and see what the Travis logs reveal?

**Comment by GeorgeLuo** [comment]
I can do that. here is what I'm observing

`[01:23:34] Starting 'validator-java'...
install_bazel.sh Bazel binary detected at bazel-installer/bazel-3.0.0-installer-darwin-x86_64.sh; skipping installation
Both validator.proto and validator-all.protoascii exist
Copying Validator.java from /Users/gluo17/Documents/workspace/amphtml/validator/java/bazel-bin/amp/validator
done copying
`

I added the "done copying" to the end of [amphtml/validator/java/src/main/bin/copyValidatorJavaSource.sh](https://github.com/ampproject/amphtml/blob/master/validator/java/src/main/bin/fetchAMPResources.sh)

It hangs after that. So it looks like it's not returning from that script gracefully, or whatever is calling that is not returning gracefully. What is the path from "gulp validator-java" to the shell script? I see it referenced in BUILD and nowhere else. But I'm not sure who is handling that BUILD file. Is it bazel or is maven being invoked?

**Comment by GeorgeLuo** [comment]
When I run the [build and test](https://github.com/ampproject/amphtml/blob/master/validator/java/build_and_test.sh) script, the ./build.sh step takes 30 seconds or so, successfully though. Could there be some kind of timeout when bringing travis into the equation?

**Comment by GeorgeLuo** [comment]
It worked courtesy of @nhant01. bazel was silently failing during the build phase because we only updated the tagchowder library in maven and didn't do the same in the bazel WORKSPACE. we'll probably need to figure out how to splay these types of errors. I'm going to clean up some of the echos and such.

**Comment by rsimha** [comment]
@GeorgeLuo @nhant01 These tests are still failing during CI. See [logs](https://travis-ci.org/github/ampproject/amphtml/jobs/677960140#L375-L550). Any idea what's going on?

**Comment by rsimha** [comment]
These tests are still failing during CI. See [logs](https://travis-ci.org/github/ampproject/amphtml/jobs/677960140#L375-L550). Any idea what's going on?

/invite @GeorgeLuo @nhant01

**Comment by GeorgeLuo** [comment]
unclear but pulling the latest proto seems to have introduced compilation failure cases to do with css specs. I imagine we're in a state where the .proto under java validator compiled fine (outdated), but the protoascii has introduced new models and flows. I'll get the engine changes in and then see about the tests.

**Comment by rsimha** [comment]
Thanks @GeorgeLuo. Meanwhile, we have a [large backlog](https://github.com/ampproject/amphtml/pulls?q=is%3Apr+is%3Aopen+%22validator%2Fjava%2Fpom.xml%22+sort%3Aupdated-desc) of package upgrade PRs for the Java validator that are all blocked by the test failures. Since `validator/java/pom.xml` currently has several outdated dependencies, would it make sense for you to do a one-time upgrade of all of them?

Under normal circumstances, once the file is up to date, there will be far fewer upgrade PRs each week, and they can be individually tested and merged as they appear.

**Pull request merged**